internal/graph/track_check_resolver.go

```diff
@@ -0,0 +1,211 @@
+
+func WithTrackerLogger(logger logger.Logger) TrackerCheckResolverOpt {
+	return func(t *TrackerCheckResolver) {
+		t.logger = logger
+	}
+}
+
+func WithTrackerContext(ctx context.Context) TrackerCheckResolverOpt {
+	return func(t *TrackerCheckResolver) {
+		t.ctx = ctx
+	}
+}
+
+// Expired check the current tuple entry expiration.
+func (r *resolutionTree) expired() bool {
+	return time.Since(r.tm) > trackerInterval
+}
+
+// NewTrackCheckResolver creates an instance tracker Resolver.
+func NewTrackCheckResolver(opts ...TrackerCheckResolverOpt) *TrackerCheckResolver {
+	t := &TrackerCheckResolver{
+		ticker: time.NewTicker(trackerInterval),
+	}
+
+	for _, opt := range opts {
+		opt(t)
+	}
+
+	t.delegate = t
+	t.launchFlush()
+	return t
+}
+
+// LogExecutionPaths reports the model and tuple path.
+func (t *TrackerCheckResolver) logExecutionPaths() {
+	t.nodes.Range(func(k, v any) bool {
+		modelid := k.(string)
+		paths, _ := v.(*sync.Map)
+		paths.Range(func(k, v any) bool {
+			tree, _ := v.(*resolutionTree)
+			path := k.(string)
+			if tree.expired() {
+				t.logger.Info("execution path hits",
+					zap.String("modelid", modelid),
+					zap.String("path", path),
+					zap.Uint64("hits", tree.hits.Load()))
+				paths.Delete(path)
+			}
+			return true
+		})
+		return true
+	})
+}
+
+// LaunchFlush starts the execution path logging and removal of entries.
+func (t *TrackerCheckResolver) launchFlush() {
+	go func() {
+		for {
+			select {
+			case <-t.ctx.Done():
+				t.ticker.Stop()
+				return
+			case <-t.ticker.C:
+				t.logExecutionPaths()
+				t.ticker.Reset(trackerInterval)
+			}
+		}
+	}()
+}
+
+// SetDelete assigns the next deletegate in the chain.
+func (t *TrackerCheckResolver) SetDelegate(delegate CheckResolver) {
+	t.delegate = delegate
+}
+
+// GetDelegate return the assigned delegate.
+func (t *TrackerCheckResolver) GetDelegate() CheckResolver {
+	return t.delegate
+}
+
+// Close implements CheckResolver.
+func (*TrackerCheckResolver) Close() {}
+
+// UserType returns the associated tuple user type.
+func (t *TrackerCheckResolver) userType(userKey string) string {
+	userObj, userRel := tuple.SplitObjectRelation(userKey)
+	_, userObjID := tuple.SplitObject(userObj)
+
+	if userRel == "" && userObjID == "*" {
+		return string(tuple.UserSet)
+	}
+
+	if userRel == "" {
+		return string(tuple.User)
+	}
+
+	return string(tuple.UserSet)
+}
+
+// GetTK returns formatted tuple suitable insertion into list.
+func (t *TrackerCheckResolver) getTK(tk *openfgav1.TupleKey) string {
+	return fmt.Sprintf("%s#%s#%s", t.userType(tk.GetUser()), tk.GetRelation(), tk.GetObject())
+}
+
+// LoadModel populate model id for individual tuple paths.
+func (t *TrackerCheckResolver) loadModel(r *ResolveCheckRequest) (value any, ok bool) {
+	model := r.GetAuthorizationModelID()
+	value, ok = t.nodes.Load(model)
+	if !ok {
+		value = &sync.Map{}
+		value.(*sync.Map).Store(t.getTK(r.GetTupleKey()), &resolutionTree{tm: time.Now(), hits: &atomic.Uint64{}})
```

Comment:
If its a sync.Map, does it also need to be atomic?

Reply:
sync.map provides low-level primitives needed to support concurrency.

Reply:
I was referring to the need to use atomic for the hits since you already had sync.Map.

Reply:
Yes it is needed. During testing I ran into concurrency issues and to use atomic add to resolve it.